velox/functions/lib/Re2Functions.cpp

```diff
@@ -1193,6 +1203,370 @@ std::shared_ptr<exec::VectorFunction> makeRe2MatchImpl(
+    }
+    if (patterns->mayHaveNulls()) {
+      nonNullRows->deselectNulls(patterns->nulls(), 0, rows.end());
+    }
+
+    auto it = transforms->iterator(nonNullRows.get());
+    while (auto entry = it.next()) {
+      // For each input row, apply regex pattern to the string and identify
+      // matches and matching groups. Total number of matches may be less or
+      // more than number of input rows. Some strings may have no matches, some
+      // may have more than one match.
+      exec::LocalSelectivityVector remainingRows(context, *entry.rows);
+      const auto matches =
+          findMatches(*strings, *patterns, *remainingRows, context);
+
+      // Exclude rows with invalid regex from further processing.
+      context.deselectErrors(*remainingRows);
+
+      // Total number of matches across all input rows. Each match has a list of
+      // matching groups.
+      const auto totalMatches = matches.matches->size();
+
+      // Number of matches per input row.
+      auto* rawNumMatches = matches.rawNumMatches;
+
+      // If lambda is using captures, make sure to 'replicate' captures for all
+      // matches of a given input row.
+      BufferPtr wrapCapture = makeWrapCapture(
+          *remainingRows, totalMatches, rawNumMatches, context.pool());
+
+      // Lambda is applied to an ArrayVector, where each row contains a list of
+      // matching groups from a single match.
+      const std::vector<VectorPtr> lambdaArgs = {matches.matches};
+
+      // Lambda function takes an array of matching groups and returns a single
+      // replacement string.
+      auto replacements =
+          BaseVector::create(VARCHAR(), totalMatches, context.pool());
+
+      SelectivityVector matchesRows(totalMatches);
+
+      entry.callable->apply(
+          matchesRows,
+          nullptr, // No need to preserve any values in 'replacements'.
+          wrapCapture,
+          &context,
+          lambdaArgs,
+          wrapCapture,
+          &replacements);
+
+      exec::LocalDecodedVector decodedReplacements(
+          context, *replacements, matchesRows);
+      auto* rawOffsets = matches.rawOffsets;
+      auto* rawSizes = matches.rawSizes;
+
+      vector_size_t matchRow = 0;
+      remainingRows->applyToSelected([&](auto row) {
+        const auto original = strings->valueAt<StringView>(row);
+
+        bool resultIsNull = false;
+
+        // Calculate string length after replacing all matches.
+        size_t resultSize = original.size();
+        for (auto i = 0; i < rawNumMatches[row]; ++i) {
+          if (decodedReplacements->isNullAt(matchRow + i)) {
+            // If one of the replacements is NULL, the result is NULL.
+            resultIsNull = true;
+            break;
+          }
+
+          const auto replacementSize =
+              decodedReplacements->valueAt<StringView>(matchRow + i).size();
+
+          resultSize -= rawSizes[matchRow + i];
+          resultSize += replacementSize;
+        }
```

Comment:
Do we need to calculate resultSize upfront? I wonder whether we can update the result size while doing the replacement through StringWriter::append() to avoid this separate loop. This could cause the string buffer to be reallocated when the capacity is not enough, but since we allocate at least 32K string buffer at a time, reallocation shouldn't be frequent I think.

Reply:
I think we can make it work with StringWriter::append(). However, using append involves checking 'newSize' against capacity and updating the 'size'. When there are many matches and replacements that could add unnecessary overhead. Also, if some replacements are NULL, we'll have to abandon partially written string. What benefits do you see of switching to StringWriter::append?

Reply:
The suggestion was to make the code cleaner. But if it affects the performance, then it's fine to keep the current code. Checking the newSize against capacity shouldn't be a problem since we're already using an additional loop here, but abandoning partially written string could be. So I think it's good to keep the current code.